File: quicksort/quicksort.py
from random import randint

def partition(arr, low, high):
	'''Puts all elements of arr smaller than arr[high-1]
	to the left and all the larger to the right of the array
	'''
	pivot = arr[high]
	i = low -1 #Begins at 0

	for j in range(low, high):
		if arr[j] < pivot:
			i+=1
			arr[i], arr[j] = arr[j], arr[i]

	arr[i+1], arr[high] = arr[high], arr[i+1]

	return i+1

def quicksort(arr, low, high):
	if low < high:

		part = partition(arr, low, high)
		quicksort(arr, low, part-1)
		quicksort(arr, part+1, high)





if __name__ == '__main__':

	# Testing array
	to_sort = [10, 5, 3, 20, 4, 50]

	quicksort(to_sort,0, len(to_sort)-1)
	print(to_sort)

	# Tesitng random array
	to_sort = [randint(0,100) for i in range(200)]
	print("array to sort:\n{}".format(to_sort))

	quicksort(to_sort, 0, len(to_sort)-1)
	print("sorted array:\n{}".format(to_sort))

File: quicksort/quicksort.c

#include <stdio.h>
#include <stdlib.h>

void swap(int* a, int* b){
	int a_val = *a;

	*a = *b;
	*b = a_val;
}

void printArray(int arr[], int size)  
{  
    int i;  
    for (i = 0; i < size; i++){
    	printf("%d ", arr[i]);
    }  
    printf("\n");
}

int partition(int arr[], int low, int high){
	int i = low-1; //starting at -1
	int pivot = arr[high]; //starting at lenght -1 (last element)

	for(int j=low; j<=high-1; j++){
		if(arr[j] < pivot){
			i++;
			swap(&arr[i], &arr[j]);
		}
	}
	swap(&arr[i+1], &arr[high]);
	return(i+1);
}

void quicksort(int arr[], int low, int high){
	if(low < high){
		int mid = partition(arr, low, high);

		quicksort(arr, low, mid-1);
		quicksort(arr, mid+1, high);

	}
}

int main(void){

	//int size = 7;
	//int array[] = {10, 9, 20, 5, 3, 13, 7};

	int size = 7;
	int *arr = (int*)malloc(size*sizeof(int));

	arr[0] = 10;
	arr[1] = 9;
	arr[2] = 20;
	arr[3] = 5;
	arr[4] = 3;
	arr[5] = 13;
	arr[6] = 7;
    
    printf("Initial array: \n"); 
    printArray(arr, size);

    quicksort(arr, 0, size-1); 

    printf("Sorted array: \n"); 
    printArray(arr, size); 

	return 0;
}
